File: libsignal-service/src/test/java/org/whispersystems/signalservice/api/groupsv2/GroupsV2Operations_ban_Test.java
package org.whispersystems.signalservice.api.groupsv2;

import org.junit.Before;
import org.junit.Test;
import org.signal.libsignal.zkgroup.InvalidInputException;
import org.signal.libsignal.zkgroup.groups.GroupMasterKey;
import org.signal.libsignal.zkgroup.groups.GroupSecretParams;
import org.signal.storageservice.protos.groups.GroupChange;
import org.signal.storageservice.protos.groups.local.DecryptedBannedMember;
import org.whispersystems.signalservice.api.push.ServiceId;
import org.whispersystems.signalservice.api.push.ServiceId.ACI;
import org.whispersystems.signalservice.internal.util.Util;
import org.whispersystems.signalservice.testutil.LibSignalLibraryUtil;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

import okio.ByteString;

import static org.hamcrest.CoreMatchers.hasItems;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.whispersystems.signalservice.api.groupsv2.ProtoTestUtils.bannedMember;

public final class GroupsV2Operations_ban_Test {

  private GroupsV2Operations.GroupOperations groupOperations;

  @Before
  public void setup() throws InvalidInputException {
    LibSignalLibraryUtil.assumeLibSignalSupportedOnOS();

    TestZkGroupServer  server             = new TestZkGroupServer();
    GroupSecretParams  groupSecretParams  = GroupSecretParams.deriveFromMasterKey(new GroupMasterKey(Util.getSecretBytes(32)));
    ClientZkOperations clientZkOperations = new ClientZkOperations(server.getServerPublicParams());

    groupOperations = new GroupsV2Operations(clientZkOperations, 10).forGroup(groupSecretParams);
  }

  @Test
  public void addBanToEmptyList() {
    ACI ban = ACI.from(UUID.randomUUID());

    GroupChange.Actions.Builder banUuidsChange = groupOperations.createBanServiceIdsChange(Collections.singleton(ban),
                                                                                           false,
                                                                                           Collections.emptyList());

    assertThat(banUuidsChange.addBannedMembers.size(), is(1));
    assertThat(banUuidsChange.addBannedMembers.get(0).added.userId, is(groupOperations.encryptServiceId(ban)));
  }

  @Test
  public void addBanToPartialFullList() {
    ACI                         toBan         = ACI.from(UUID.randomUUID());
    List<DecryptedBannedMember> alreadyBanned = new ArrayList<>(5);

    for (int i = 0; i < 5; i++) {
      alreadyBanned.add(bannedMember(UUID.randomUUID()));
    }

    GroupChange.Actions.Builder banUuidsChange = groupOperations.createBanServiceIdsChange(Collections.singleton(toBan),
                                                                                           false,
                                                                                           alreadyBanned);

    assertThat(banUuidsChange.addBannedMembers.size(), is(1));
    assertThat(banUuidsChange.addBannedMembers.get(0).added.userId, is(groupOperations.encryptServiceId(toBan)));
  }

  @Test
  public void addBanToFullList() {
    ACI                         toBan         = ACI.from(UUID.randomUUID());
    List<DecryptedBannedMember> alreadyBanned = new ArrayList<>(10);
    DecryptedBannedMember       oldest        = null;

    for (int i = 0; i < 10; i++) {
      DecryptedBannedMember member = bannedMember(UUID.randomUUID()).newBuilder().timestamp(100 + i).build();
      if (oldest == null) {
        oldest = member;
      }
      alreadyBanned.add(member);
    }

    Collections.shuffle(alreadyBanned);

    GroupChange.Actions.Builder banUuidsChange = groupOperations.createBanServiceIdsChange(Collections.singleton(toBan),
                                                                                           false,
                                                                                           alreadyBanned);

    assertThat(banUuidsChange.deleteBannedMembers.size(), is(1));
    assertThat(banUuidsChange.deleteBannedMembers.get(0).deletedUserId, is(groupOperations.encryptServiceId(ServiceId.parseOrThrow(oldest.serviceIdBytes))));


    assertThat(banUuidsChange.addBannedMembers.size(), is(1));
    assertThat(banUuidsChange.addBannedMembers.get(0).added.userId, is(groupOperations.encryptServiceId(toBan)));
  }

  @Test
  public void addMultipleBanToFullList() {
    List<ACI> toBan = new ArrayList<>();
    toBan.add(ACI.from(UUID.randomUUID()));
    toBan.add(ACI.from(UUID.randomUUID()));

    List<DecryptedBannedMember> alreadyBanned = new ArrayList<>(10);
    for (int i = 0; i < 10; i++) {
      alreadyBanned.add(bannedMember(UUID.randomUUID()).newBuilder().timestamp(100 + i).build());
    }

    List<ByteString> oldest = new ArrayList<>(2);
    oldest.add(groupOperations.encryptServiceId(ServiceId.parseOrThrow(alreadyBanned.get(0).serviceIdBytes)));
    oldest.add(groupOperations.encryptServiceId(ServiceId.parseOrThrow(alreadyBanned.get(1).serviceIdBytes)));

    Collections.shuffle(alreadyBanned);

    GroupChange.Actions.Builder banUuidsChange = groupOperations.createBanServiceIdsChange(new HashSet<>(toBan),
                                                                                           false,
                                                                                           alreadyBanned);

    assertThat(banUuidsChange.deleteBannedMembers.size(), is(2));
    assertThat(banUuidsChange.deleteBannedMembers
                             .stream()
                             .map(a -> a.deletedUserId)
                             .collect(Collectors.toList()),
               hasItems(oldest.get(0), oldest.get(1)));


    assertThat(banUuidsChange.addBannedMembers.size(), is(2));
    assertThat(banUuidsChange.addBannedMembers
                             .stream()
                             .map(a -> a.added)
                             .map(b -> b.userId)
                             .collect(Collectors.toList()),
               hasItems(groupOperations.encryptServiceId(toBan.get(0)),
                        groupOperations.encryptServiceId(toBan.get(1))));
  }
}

File: libsignal-service/src/test/java/org/whispersystems/signalservice/api/groupsv2/GroupsV2Operations_ban_Test.kt
package org.whispersystems.signalservice.api.groupsv2

import assertk.assertThat
import assertk.assertions.containsExactly
import assertk.assertions.isEqualTo
import assertk.assertions.single
import org.junit.Before
import org.junit.Test
import org.signal.libsignal.zkgroup.groups.GroupMasterKey
import org.signal.libsignal.zkgroup.groups.GroupSecretParams
import org.whispersystems.signalservice.api.groupsv2.GroupsV2Operations.GroupOperations
import org.whispersystems.signalservice.api.push.ServiceId
import org.whispersystems.signalservice.internal.util.Util
import org.whispersystems.signalservice.testutil.LibSignalLibraryUtil
import java.util.UUID

@Suppress("ClassName")
class GroupsV2Operations_ban_Test {
  private lateinit var groupOperations: GroupOperations

  @Before
  fun setup() {
    LibSignalLibraryUtil.assumeLibSignalSupportedOnOS()

    val server = TestZkGroupServer()
    val groupSecretParams = GroupSecretParams.deriveFromMasterKey(GroupMasterKey(Util.getSecretBytes(32)))
    val clientZkOperations = ClientZkOperations(server.serverPublicParams)

    groupOperations = GroupsV2Operations(clientZkOperations, 10).forGroup(groupSecretParams)
  }

  @Test
  fun addBanToEmptyList() {
    val ban = randomACI()

    val banUuidsChange = groupOperations.createBanServiceIdsChange(
      /* banServiceIds = */
      setOf(ban),
      /* rejectJoinRequest = */
      false,
      /* bannedMembersList = */
      emptyList()
    )

    assertThat(banUuidsChange.addBannedMembers)
      .single()
      .transform { it.added?.userId }
      .isEqualTo(groupOperations.encryptServiceId(ban))
  }

  @Test
  fun addBanToPartialFullList() {
    val toBan = randomACI()
    val alreadyBanned = (0 until 5).map { ProtoTestUtils.bannedMember(UUID.randomUUID()) }

    val banUuidsChange = groupOperations.createBanServiceIdsChange(
      /* banServiceIds = */
      setOf(toBan),
      /* rejectJoinRequest = */
      false,
      /* bannedMembersList = */
      alreadyBanned
    )

    assertThat(banUuidsChange.addBannedMembers)
      .single()
      .transform { it.added?.userId }
      .isEqualTo(groupOperations.encryptServiceId(toBan))
  }

  @Test
  fun addBanToFullList() {
    val toBan = ServiceId.ACI.from(UUID.randomUUID())

    val alreadyBanned = (0 until 10).map { i ->
      ProtoTestUtils.bannedMember(UUID.randomUUID())
        .newBuilder()
        .timestamp(100L + i)
        .build()
    }.shuffled()

    val banUuidsChange = groupOperations.createBanServiceIdsChange(
      /* banServiceIds = */
      setOf(toBan),
      /* rejectJoinRequest = */
      false,
      /* bannedMembersList = */
      alreadyBanned
    )

    val oldest = alreadyBanned.minBy { it.timestamp }
    assertThat(banUuidsChange.deleteBannedMembers)
      .single()
      .transform { it.deletedUserId }
      .isEqualTo(groupOperations.encryptServiceId(ServiceId.parseOrThrow(oldest.serviceIdBytes)))

    assertThat(banUuidsChange.addBannedMembers)
      .single()
      .transform { it.added?.userId }
      .isEqualTo(groupOperations.encryptServiceId(toBan))
  }

  @Test
  fun addMultipleBanToFullList() {
    val toBan = (0 until 2).map { ServiceId.ACI.from(UUID.randomUUID()) }

    val alreadyBanned = (0 until 10).map { i ->
      ProtoTestUtils.bannedMember(UUID.randomUUID())
        .newBuilder()
        .timestamp(100L + i)
        .build()
    }.shuffled()

    val banUuidsChange = groupOperations.createBanServiceIdsChange(
      /* banServiceIds = */
      toBan.toMutableSet(),
      /* rejectJoinRequest = */
      false,
      /* bannedMembersList = */
      alreadyBanned
    )

    val oldestTwo = alreadyBanned
      .sortedBy { it.timestamp }
      .subList(0, 2)
      .map { groupOperations.encryptServiceId(ServiceId.parseOrThrow(it.serviceIdBytes)) }
      .toTypedArray()
    assertThat(banUuidsChange.deleteBannedMembers)
      .transform { members ->
        members.map { member ->
          member.deletedUserId
        }
      }
      .containsExactly(*oldestTwo)

    val newBans = (0..1).map { i ->
      groupOperations.encryptServiceId(toBan[i])
    }.toTypedArray()
    assertThat(banUuidsChange.addBannedMembers)
      .transform { members ->
        members.map { member ->
          member.added?.userId
        }
      }
      .containsExactly(*newBans)
  }

  private fun randomACI(): ServiceId.ACI = ServiceId.ACI.from(UUID.randomUUID())
}
